### ISSUE TITLE
Consider categorizing output of 'help' command

### ISSUE THREAD

**Issue by kotfu** [python-cmd2/cmd2#278]
Right now the help command just gives you an alphabetical columnar list of commands. It would be nice if there was a way to create groups or categories of commands, so you could have related commands show up together.

Instead of this:

```
Documented commands (type help <topic>):
========================================
config     expire       load      resources   shell                stop
connect    findleakers  py        restart     shortcuts            threaddump
deploy     help         pyscript  serverinfo  show                 undeploy
edit       history      quit      sessions    sslconnectorciphers  version
exit       license      redeploy  set         start                vminfo
exit_code  list         reload    settings    status               which
```
It would be nice if you could create something like:
```
Documented commands (for more info type help <command>)

Connecting
=========================================
connect     which

Application Management
=========================================
list
deploy       redeploy      undeploy
start        restart       stop
sessions     expire        findleakers

Server Information
==========================================
resources         serverinfo      sslconnectorciphers
status            thread dump     vminfo

Settings
==========================================
config       set           show

Other
=============================
edit      help        license
exit      history     load
version
```

It would be great if you could not only control the sections of the help, but the order and formatting of the commands within each section. Commands not specified for inclusion in a specific section would be featured in a catch all section like **Other**.

It might also be interesting to have a way to specify a short phrase of help for each command, with some command like **help --long** to show:

```
Documented commands (for more info type help <command>)

Connecting
=========================================
connect    connect to a tomcat server
which      show which server you are currently connected to
...
```

**Comment by tleonhardt** [comment]
I really like the concept of having categories of commands which get displayed together in the help menu, particularly if this is an opt-in capability which defaults to the current behavior by default.  That would be a nice addition.

Potentially of relevance, the existing behavior actually divides the help menu into three categories:
- Documented commands
    - A **do_foo** method exists which has a docstring and/or an associated **help_foo** method
- Miscellaneous
    - A **help_foo** method exists, but there is no corresponding **do_foo**
- Undocumented commands
    - A **do_foo** method exists which does not have a docstring or an associated **help_foo** method

**Comment by anselor** [comment]
I just added a low-impact implementation of help categories in the help_categories branch. 
Let me know what you guys think of this approach.

For the **help --long** (or maybe **help --verbose**), I was thinking just take the first non-empty line of function docstring used in the help for the command. Any thoughts?

**Comment by tleonhardt** [comment]
@anselor cmd2 will still be supporting Python 2.7 until the end of August 2018.  Optional type hinting is only available in Python 3.  So you will need to remove your usage of optional type hinting.

Also, there appear to be a lot of unit test failures for every version of Python other than 3.5.

We currently support versions 2.7, 3.4, 3.5, 3.6, and 3.7.

**Comment by tleonhardt** [comment]
If recommend setting up multiple versions of Python in different virtual environments for testing.  This can be done either using the **conda** package manager which comes with the Anaconda distribution or the **virtualenv** module.

**Comment by tleonhardt** [comment]
@anselor If you haven't already, check out the [Contributing Guide](https://github.com/python-cmd2/cmd2/blob/master/CONTRIBUTING.md) for advice on how to run the unit tests and such.

**Comment by anselor** [comment]
help -v / help --verbose
![screenshot_2018-04-10_14-46-05](https://user-images.githubusercontent.com/17968842/38577155-285b48a8-3cce-11e8-8dac-bdc10d7120fe.png)

**Comment by tleonhardt** [comment]
I like the verbose help where you see a brief description about what the command does.

**Comment by anselor** [comment]
I ran into a slight snag with the verbose help. 
I'm currently checking the __doc__ property of the function. In the cases where there's a help_<command> function provided, this overrides the __doc__ string. The help_<command> functions don't return a string. Instead, they directly write their output to stdout by whatever means they wish. I could temporarily redirect all of stdout to a string to capture this but I'm concerned about what side effects this could have. It could cause weird behavior in multi-threaded applications that happen to write print something while we're generating help output.

@tleonhardt @kmvanbrunt

**Comment by tleonhardt** [comment]
This was closed by #348
